stl/inc/algorithm

```diff
@@ -1865,6 +1865,199 @@ _NODISCARD _FwdIt1 find_end(_ExPo&& _Exec, _FwdIt1 _First1, _FwdIt1 _Last1, _Fwd
+
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_UNext1), _STD invoke(_Proj2, *_UNext2))) {
+                        break; // mismatch, go to the next candidate
+                    }
+
+                    ++_UNext1;
+                    ++_UNext2;
+                }
+
+                ++_UFirst1;
+            }
+        }
+
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <forward_iterator _It1, sentinel_for<_It1> _Se1, forward_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pr = ranges::equal_to, class _Pj1 = identity, class _Pj2 = identity>
+            requires indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr subrange<_It1> operator()(_It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2,
+            _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+            if constexpr (random_access_iterator<_It1> && sized_sentinel_for<_Se1, _It1>
+                && random_access_iterator<_It2> && sized_sentinel_for<_Se2, _It2>) {
+                return _Random_access_sized_ranges(_First1, _Last1 - _First1, _First2, _Last2 - _First2,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else if constexpr (bidirectional_iterator<_It1> && is_same_v<_It1, _Se1>
+                && bidirectional_iterator<_It2> && is_same_v<_It2, _Se2>) {
+                return _Bidi_common_ranges(_First1, _Last1, _First2, _Last2,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else {
+                return _Forward_ranges(_First1, _Last1, _First2, _Last2,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            }
+        }
+
+        template <forward_range _Rng1, forward_range _Rng2, class _Pr = ranges::equal_to, class _Pj1 = identity,
+            class _Pj2 = identity>
+            requires indirectly_comparable<iterator_t<_Rng1>, iterator_t<_Rng2>, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr borrowed_subrange_t<_Rng1> operator()(
+            _Rng1&& _Range1, _Rng2&& _Range2, _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            if constexpr (random_access_range<_Rng1> && sized_range<_Rng1>
+                && random_access_range<_Rng2> && sized_range<_Rng2>) {
+                return _Random_access_sized_ranges(
+                    _RANGES begin(_Range1), _RANGES distance(_Range1),
+                    _RANGES begin(_Range2), _RANGES distance(_Range2),
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else if constexpr (bidirectional_range<_Rng1> && common_range<_Rng1>
+                && bidirectional_range<_Rng2> && common_range<_Rng2>) {
+                return _Bidi_common_ranges(
+                    _RANGES begin(_Range1), _RANGES end(_Range1),

```

Comment:
why do we need this macro for the range namespace, I was under the impression STD and CSTD were mostly legacy

Reply:
`_STD` and `_CSTD` (and sometimes `_RANGES`, but not in this case) are used to fully qualify names to avoid ADL hijacking. We use `_RANGES` with function objects like this mostly to indicate that nobody forgot to fully-qualify.

Reply:
right I mean as opposed to just writing ::std::ranges:: or ::std::

Reply:
Filed #825 to discuss replacing these macros with their expansions.